### PR TITLE
Display python error message when kernel dies

### LINKED ISSUE #7318: Provide a better error message when kernel dies

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #7318

### CODE REVIEW

File: src/client/datascience/jupyter/kernels/kernelCommandListener.ts [hunk 1]
```diff
@@ -208,9 +208,14 @@ export class KernelCommandListener implements IDataScienceCommandListener {
...
             } else {
-                // Show the error message
-                void this.applicationShell.showErrorMessage(exc);
-                traceError(exc);
+                traceError('Failed to restart the kernel', exc);
+                if (exc) {
+                    // Show the error message
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    void this.applicationShell.showErrorMessage(
+                        exc instanceof Error ? exc.message : (exc as any).toString()
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelCommandListener.ts:216]: Cleaner than displaying a to string version of the error.

File: src/client/datascience/jupyter/kernels/kernelExecution.ts [hunk 2]
```diff
@@ -245,9 +245,7 @@ export class KernelExecution implements IDisposable {
     @captureTelemetry(Telemetry.RestartJupyterTime)
     private async restartExecution(notebook: INotebook): Promise<void> {
         // Just use the internal session. Pending cells should have been canceled by the caller
-        await notebook.session.restart(this.interruptTimeout).catch((exc) => {
-            traceWarning(`Error during restart: ${exc}`);
-        });
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelExecution.ts:250]: This was a bug, we were swalling errors when restarting kernels, its as though everything went fine, when infact the kernel actually died.

File: src/client/datascience/kernel-launcher/types.ts [hunk 3]
```diff
@@ -110,9 +110,13 @@ export class PythonKernelDiedError extends BaseError {
...
     public readonly reason?: string;
     constructor(options: { exitCode: number; reason?: string; stdErr: string } | { error: Error; stdErr: string }) {
+        // Last line in stack traces generally contains the error message.
+        // Display that in the error message.
+        let reason = ('reason' in options ? options.reason || '' : options.stdErr).trim().split('\n').reverse()[0];
+        reason = reason ? `${reason}, \n` : '';
         const message =
             'exitCode' in options
-                ? `Kernel died with exit code ${options.exitCode}. ${options.reason}`
+                ? `Kernel died (code: ${options.exitCode}). ${reason}${options.reason}`
```
Comment (DonJayamanne) [inline src/client/datascience/kernel-launcher/types.ts:119]: Updated to incldue error message in the first line (in python stack traces the error message is at the end, this way we extract the last line & put it on top of the error message)

Comment (rchiodo) [reply src/client/datascience/kernel-launcher/types.ts:119]: What does it look like in the UI?

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/kernel-launcher/types.ts (hunk 3 above)
Comment (DonJayamanne) [reply src/client/datascience/kernel-launcher/types.ts:119]: Sorry, meant to upload that, took the screenshots earlier.
here they are
Screen Shot 2021-09-23 at 14 42 11
Screen Shot 2021-09-23 at 14 50 40

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**